addons/sale_timesheet/models/sale_order.py

```diff
@@ -48,17 +48,38 @@ def _compute_field_value(self, field):
             return
 
         # Get SOs which their state is not equal to upselling or invoied and if at least a SOL has warning prepaid service upsell set to True and the warning has not already been displayed
-        filtered_self = self.filtered(lambda so: (so.user_id or so.partner_id.user_id) and so.invoice_status not in ('upselling', 'invoiced') and any(sol._has_warning_prepaid_service_upsell() and not sol.has_displayed_warning_upsell for sol in so.order_line))
-        if not filtered_self:
-            return
-        warning_upsell_sols = filtered_self.order_line._get_warning_prepaid_service_upsell_sols()
-        filtered_self = warning_upsell_sols.order_id
-        if not filtered_self:
-            return
-
-        filtered_self._create_upsell_activity()
-        # We want to display only one time the warning for each SOL
-        warning_upsell_sols.write({'has_displayed_warning_upsell': True})
+        upsellable_orders = self.filtered(lambda so:
+            so.state == 'sale'
+            and so.invoice_status not in ('upselling', 'invoiced')
+            and (so.user_id or so.partner_id.user_id)  # salesperson needed to assign upsell activity
+        )
+        for order in upsellable_orders:
+            upsellable_lines = order._filter_prepaid_service_lines_to_upsell()
+            if upsellable_lines:
+                order._create_upsell_activity()
+                # We want to display only one time the warning for each SOL
+                upsellable_lines.write({'has_displayed_warning_upsell': True})
+
+    def _filter_prepaid_service_lines_to_upsell(self):
+        """ Retrieve all sols which need to display an upsell activity warning in the SO
+
+            These SOLs should contain a product which has:
+                - type="service",
+                - service_policy="ordered_timesheet",
+                - service_upsell_warning=True.
+        """
+        self.ensure_one()
+        precision = self.env['decimal.precision'].precision_get('Product Unit of Measure')
+        return self.order_line.filtered(lambda sol:
+            not sol.has_displayed_warning_upsell  # we don't want to display many times the warning each time we timesheet on the SOL
+            and sol._has_warning_prepaid_service_upsell()
```

Comment:
Since `_has_warning_prepaid_service_upsell` is only used here and has no override, you can move the conditions directly into the lambda here ;).

Reply:
It's true and it is done :) 
I remove the following condition `sol.product_id.invoice_policy == 'order'` because sol.product_id.service_policy = 'ordered_timesheet' if `sol.product_id.invoice_policy == 'order' and (sol.product_id.service_type == 'timesheet' or sol.product_id.type == 'service')`. Thus, this condition was already checked in compute of `service_type` field.
You can see it in this [link](https://github.com/odoo/odoo/blob/15cee18b92f80b8b5919d03bbd2e8a58e0a1e7f1/addons/sale_timesheet/models/product.py#L44).